LibreNMS/OS/Teldat.php

```diff
@@ -0,0 +1,139 @@
+<?php
+/**
+ * Teldat.php
+ *
+ * Teldat Devices
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ *
+ * @link       https://www.librenms.org
+ *
+ * @copyright  2021 Antonio Pedro Santos
+ * @author     Antonio Pedro Santos <cupidazul@gmail.com>
+ */
+
+namespace LibreNMS\OS;
+
+use App\Models\Port;
+use Illuminate\Support\Str;
+use LibreNMS\Device\WirelessSensor;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessCellDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessClientsDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRsrpDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRsrqDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRssiDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessSinrDiscovery;
+use LibreNMS\OS;
+use LibreNMS\OS\Traits\TeldatCellular;
+
+class Teldat extends OS implements
+    WirelessCellDiscovery,
+    WirelessClientsDiscovery,
+    WirelessRssiDiscovery,
+    WirelessRsrqDiscovery,
+    WirelessRsrpDiscovery,
+    WirelessSinrDiscovery
+{
+    use TeldatCellular;
+
+    private $deviceInfo = [];
+    private $hasWLANif = false;
+    private $hasGSMif = false;
+
+    public function getdevicePortsInfo()
+    {
+        $this->deviceInfo = Port::whereIn('device_id', [$this->getDeviceId()])
```

Comment:
Database will not have been populated when this discovery is first run resulting in erroneous data.
Please use snmp data to determine if the interfaces exist.

```php
$ifNames = $this->getCacheByIndex('ifName', 'IF-MIB');
```

Also, this should probably be two separate functions and return a boolean.

Reply:
The idea was to avoid recurrent requests to the if-mib, during tests I saw that there were several requests to the if-mib.
During my tests I found no issues during discovery with the procedure here in place.

Reply:
You are using getCacheByIndex wrong.  You need to call it on ifName, not ifName.2 (or whatever)

Reply:
Ok I'll try that and do some more testing on my side. Thx.

Reply:
I've done some tests, here are my findings:

I've replaced  " $this->deviceInfo = Port::whereIn('device_id', [$this->getDeviceId()]) " with "getCacheByIndex"  :

```
    public function getdevicePortsInfo()
    {

        $this->deviceIfInfo['ifName'] = $this->getCacheByIndex('ifName', 'IF-MIB');
        $this->deviceIfInfo['ifOperStatus'] = $this->getCacheByIndex('ifOperStatus', 'IF-MIB');

        foreach ($this->deviceIfInfo['ifName'] as $entIndex => $entValue) {

            // Check if any wlan interface is up
            if (Str::startsWith($entValue, 'wlan') and $this->deviceIfInfo['ifOperStatus'][$entIndex] == 'up') {
                $this->hasWLANif = true;
            }

            // Check if any cellular interface is up
            if (Str::startsWith($entValue, 'cellular') and $this->deviceIfInfo['ifOperStatus'][$entIndex] == 'up') {
                $this->hasGSMif = true;
            }

        }

    }

```

During discovery we can see that Snmp-Bulk-Walk is called first in "#### Load disco module ports ####", and a second time inside "#### Load disco module wireless ####" ...

```
#### Load disco module ports ####
.
.
.
SNMP['/usr/bin/snmpbulkwalk' '-v2c' '-c' 'PT-pwd4SNMP' '-OQUs' '-m' 'IF-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/teldat' 'udp:172.23.3.139:161' 'ifOperStatus']
ifOperStatus.1 = up
ifOperStatus.2 = up
ifOperStatus.3 = down
ifOperStatus.4 = down
ifOperStatus.5 = down
ifOperStatus.6 = up
ifOperStatus.7 = up
ifOperStatus.8 = down
ifOperStatus.9 = down
ifOperStatus.10 = lowerLayerDown
ifOperStatus.11 = lowerLayerDown
ifOperStatus.12 = up
ifOperStatus.13 = up
ifOperStatus.14 = up
ifOperStatus.15 = down
ifOperStatus.16 = down
ifOperStatus.64536 = up
ifOperStatus.64537 = up
ifOperStatus.64538 = down
ifOperStatus.64539 = down
ifOperStatus.64540 = down

array (
  1 =>
  array (
    'ifDescr' => 'ethernet0/0',
    'ifName' => 'ethernet0/0',
    'ifAlias' => '',
    'ifType' => 'iso88023Csmacd',
    'ifOperStatus' => 'up',
  ),
  2 =>
  array (
    'ifDescr' => 'ethernet0/1',
    'ifName' => 'ethernet0/1',
    'ifAlias' => '',
    'ifType' => 'iso88023Csmacd',
    'ifOperStatus' => 'up',
  ),
  3 =>
  array (
    'ifDescr' => 'cellular0/0',
    'ifName' => 'cellular0/0',
    'ifAlias' => '',
    'ifType' => 'wwanPP',
    'ifOperStatus' => 'down',
  ),
  4 =>
  array (
    'ifDescr' => 'cellular0/1',
    'ifName' => 'cellular0/1',
    'ifAlias' => '',
    'ifType' => 'ipForward',
    'ifOperStatus' => 'down',
  ),
  5 =>
  array (
    'ifDescr' => 'cellular0/2',
    'ifName' => 'cellular0/2',
    'ifAlias' => '',
    'ifType' => 'ipForward',
    'ifOperStatus' => 'down',
  ),
  6 =>
  array (
    'ifDescr' => 'cellular1/0',
    'ifName' => 'cellular1/0',
    'ifAlias' => '',
    'ifType' => 'wwanPP',
    'ifOperStatus' => 'up',
  ),
  7 =>
  array (
    'ifDescr' => 'cellular1/1',
    'ifName' => 'cellular1/1',
    'ifAlias' => '',
    'ifType' => 'ipForward',
    'ifOperStatus' => 'up',
  ),
  8 =>
  array (
    'ifDescr' => 'cellular1/2',
    'ifName' => 'cellular1/2',
    'ifAlias' => '',
    'ifType' => 'ipForward',
    'ifOperStatus' => 'down',
  ),
  9 =>
  array (
    'ifDescr' => 'wlan0/0',
    'ifName' => 'wlan0/0',
    'ifAlias' => '',
    'ifType' => 'ieee80211',
    'ifOperStatus' => 'down',
  ),
  10 =>
  array (
    'ifDescr' => 'ppp30',
    'ifName' => 'ppp30',
    'ifAlias' => '',
    'ifType' => 'ppp',
    'ifOperStatus' => 'lowerLayerDown',
  ),
  11 =>
  array (
    'ifDescr' => 'ppp40',
    'ifName' => 'ppp40',
    'ifAlias' => '',
    'ifType' => 'ppp',
    'ifOperStatus' => 'lowerLayerDown',
  ),
  12 =>
  array (
    'ifDescr' => 'direct-ip1',
    'ifName' => 'direct-ip1',
    'ifAlias' => '',
    'ifType' => 'propPointToPointSerial',
    'ifOperStatus' => 'up',
  ),
  13 =>
  array (
    'ifDescr' => 'ethernet0/1.21',
    'ifName' => 'ethernet0/1.21',
    'ifAlias' => '',
    'ifType' => 'l2vlan',
    'ifOperStatus' => 'up',
  ),
  14 =>
  array (
    'ifDescr' => 'ethernet0/0.300',
    'ifName' => 'ethernet0/0.300',
    'ifAlias' => '== LAN: UTILIZADORES ==',
    'ifType' => 'l2vlan',
    'ifOperStatus' => 'up',
  ),
  15 =>
  array (
    'ifDescr' => 'l2tp30',
    'ifName' => 'l2tp30',
    'ifAlias' => '',
    'ifType' => 'propVirtual',
    'ifOperStatus' => 'down',
  ),
  16 =>
  array (
    'ifDescr' => 'l2tp40',
    'ifName' => 'l2tp40',
    'ifAlias' => '',
    'ifType' => 'propVirtual',
    'ifOperStatus' => 'down',
  ),
  64536 =>
  array (
    'ifDescr' => 'Loopback interface',
    'ifName' => 'Loopback Interface',
    'ifAlias' => '',
    'ifType' => 'softwareLoopback',
    'ifOperStatus' => 'up',
  ),
  64537 =>
  array (
    'ifDescr' => 'Switch Port interface: Group 1, Port 1; Associated interface: ethernet0/0',
    'ifName' => 'ethernet0/0',
    'ifAlias' => '',
    'ifType' => 'iso88023Csmacd',
    'ifOperStatus' => 'up',
  ),
  64538 =>
  array (
    'ifDescr' => 'Switch Port interface: Group 1, Port 2; Associated interface: ethernet0/0',
    'ifName' => 'ethernet0/0',
    'ifAlias' => '',
    'ifType' => 'iso88023Csmacd',
    'ifOperStatus' => 'down',
  ),
  64539 =>
  array (
    'ifDescr' => 'Switch Port interface: Group 1, Port 3; Associated interface: ethernet0/0',
    'ifName' => 'ethernet0/0',
    'ifAlias' => '',
    'ifType' => 'iso88023Csmacd',
    'ifOperStatus' => 'down',
  ),
  64540 =>
  array (
    'ifDescr' => 'Switch Port interface: Group 1, Port 4; Associated interface: ethernet0/0',
    'ifName' => 'ethernet0/0',
    'ifAlias' => '',
    'ifType' => 'iso88023Csmacd',
    'ifOperStatus' => 'down',
  ),
)

.
.
.

#### Load disco module wireless ####
SQL[SELECT * FROM `wireless_sensors` WHERE `device_id`=? AND `sensor_class`=? [75,"ap-count"] 16.6ms]

clients: SNMP['/usr/bin/snmpbulkwalk' '-v2c' '-c' 'PT-pwd4SNMP' '-OQUs' '-m' 'IF-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/teldat' 'udp:172.23.3.139:161' 'ifName']
ifName.1 = ethernet0/0
ifName.2 = ethernet0/1
ifName.3 = cellular0/0
ifName.4 = cellular0/1
ifName.5 = cellular0/2
ifName.6 = cellular1/0
ifName.7 = cellular1/1
ifName.8 = cellular1/2
ifName.9 = wlan0/0
ifName.10 = ppp30
ifName.11 = ppp40
ifName.12 = direct-ip1
ifName.13 = ethernet0/1.21
ifName.14 = ethernet0/0.300
ifName.15 = l2tp30
ifName.16 = l2tp40
ifName.64536 = Loopback Interface
ifName.64537 = ethernet0/0
ifName.64538 = ethernet0/0
ifName.64539 = ethernet0/0
ifName.64540 = ethernet0/0

SNMP['/usr/bin/snmpbulkwalk' '-v2c' '-c' 'PT-pwd4SNMP' '-OQUs' '-m' 'IF-MIB' '-M' '/opt/librenms/mibs:/opt/librenms/mibs/teldat' 'udp:172.23.3.139:161' 'ifOperStatus']
ifOperStatus.1 = up
ifOperStatus.2 = up
ifOperStatus.3 = down
ifOperStatus.4 = down
ifOperStatus.5 = down
ifOperStatus.6 = up
ifOperStatus.7 = up
ifOperStatus.8 = down
ifOperStatus.9 = down
ifOperStatus.10 = lowerLayerDown
ifOperStatus.11 = lowerLayerDown
ifOperStatus.12 = up
ifOperStatus.13 = up
ifOperStatus.14 = up
ifOperStatus.15 = down
ifOperStatus.16 = down
ifOperStatus.64536 = up
ifOperStatus.64537 = up
ifOperStatus.64538 = down
ifOperStatus.64539 = down
ifOperStatus.64540 = down

```


The first Call "#### Load disco module ports ####", results in a more compreensive array byIndex:

 [index] =>
  array (
    'ifDescr' => 'ethernet0/0',
    'ifName' => 'ethernet0/0',
    'ifAlias' => '',
    'ifType' => 'iso88023Csmacd',
    'ifOperStatus' => 'up',
  ),

Shouldn't we try to use this information to avoid queying the device a second time?
And if Ports information doesn't exist, then we can call "getCacheByIndex", but we need to "reproduce" a same type of array byIndex.

Thx.